src/os/bluestore/BlueStore.cc

```diff
@@ -11776,11 +11841,134 @@ int BlueStore::read(
+
+    if (span_stat.cached == 0 && span_stat.allocated_shared == 0) {
+      if (opt_defragment || opt_recompress) {
+        // will probably need write context
+        _choose_write_options(c, o, op_flags, &wctx);
+      }
+      // do reformat if
+      // - object isn't cached (meaning it's not being written at the moment),
+      // - and there are no shared blobs withing the span.
+      //   as this might result is used space increase.
+      if (opt_recompress) {
+	if (wctx.compress &&
+	    span_stat.allocated > 0) {
+	  ceph_assert(wctx.compressor);
+	  logger->inc(l_bluestore_reformat_compress_attempted);
+	  need = 0;
+	  auto bl_it = bl.begin();
+	  uint64_t offs = offset;
+	  while(bl_it != bl.end()) {
+	    BlobRef blob = c->new_blob();
+
+	    bufferlist from_bl;
+	    size_t l = std::min(wctx.target_blob_size, uint64_t(bl_it.get_remaining()));
+	    bl_it.copy(l, from_bl);
+	    //FIXME: add zero detection
+	    auto& wi = wctx.write(offs, blob, l, 0, from_bl, 0, l, false, true);
+
+	    uint64_t res_len = from_bl.length();
+	    std::optional<int32_t> compressor_message;
+	    bufferlist cbl;
+	    if (l > min_alloc_size &&
+	        wctx.compressor->compress(from_bl, cbl, compressor_message) == 0) {
+	      res_len = cbl.length();
+	      wi.compressor_message = compressor_message;
+	      std::swap(wi.compressed_bl, cbl);
+	      // don't set wi.compress_len and wi.compressed as this is s redundant
+	      // at this point, to be assigned in _do_alloc_write if needed.
+	    }
+	    need += p2roundup(res_len, min_alloc_size);
+	    offs += l;
+	    dout(20) << __func__ << " precompress : 0x"
+	      << std::hex << offs << "~" << l << "->" << res_len
+	      << std::dec << " " << * blob
+	      << dendl;
+	  }
+	  uint64_t allocated = span_stat.allocated + span_stat.allocated_compressed;
+	  dout(10) << __func__
+	    << " recompress info, 0x"
+	    << std::hex << need << " vs. 0x" << allocated
+	    << std::dec << dendl;
+	  if (need < allocated) {
+	    might_need_recompress = true;
+	    wctx.precompressed = true;
+	  } else if (need == allocated) {
+	    wctx.precompressed = true; // may be do compression if defragmenting below
+	  } else {
+	    // don't even try compression if defragmenting below
+	    logger->inc(l_bluestore_reformat_compress_omitted);
+	    wctx.reset();
+	    need = length;
+	  }
+	}
+      }
+      if (!might_need_recompress && opt_defragment) {
+	if (span_stat.frags > 1) {
+	  logger->inc(l_bluestore_reformat_defragment_attempted);
+	  PExtentVector prealloc;
+	  prealloc.reserve(need / min_alloc_size + 1);
+	  auto start = mono_clock::now();
+	  int64_t preallocated = alloc->allocate(
+	    need, min_alloc_size, need,
+	    0, &prealloc);
+	  log_latency("allocator@_prepare_reformat",
+	    l_bluestore_allocator_lat,
+	    mono_clock::now() - start,
+	    cct->_conf->bluestore_log_op_age);
+	  might_need_defragment = preallocated >= (int64_t)need && prealloc.size() < span_stat.frags;
```

Comment:
We need to create diligent testing against all allocators we have.
My doubt is that we compare fragments in object with fragments in free space.
My thinking is - if I optimized object 5 chunks -> 2 chunks, then I probably created 5 small free chunks and removed 2 larger free chunks. There might be an emergent pattern that ultimately keeps both objects and free space more concise, but its impossible to infer it from above.

Reply:
One should realize that this PR is intended primarily for object optimization not a whole free space one. So yes hopefully object defragmentation could result in free space defragmentation as well but generally speaking that's rather a different task.